src/rm_cms.c

```diff
@@ -0,0 +1,309 @@
+    return strncasecmp(cRedisStr, str, len) == 0;
+}
+
+static int GetCMS(RedisModuleCtx *ctx, RedisModuleString *keyName, 
+                    CMSketch **cms, int mode) {
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, mode);
+    if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+        RedisModule_ReplyWithError(ctx, "CMS: key does not exist");
+        return REDISMODULE_ERR;
+    } else if(RedisModule_ModuleTypeGetType(key) != CMSketchType) {
+        RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
+        return REDISMODULE_ERR;
+    } 
+    *cms = RedisModule_ModuleTypeGetValue(key);
+    return REDISMODULE_OK;
+}
+
+static int CreateCmsKey(RedisModuleCtx *ctx, RedisModuleString *keyName, 
+        long long width, long long depth, CMSketch **cms, RedisModuleKey **key) {
+    if(*key == NULL) {
+        *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ|REDISMODULE_WRITE);
+    }
+
+    RedisModule_RetainString(ctx, keyName);
+    *cms = NewCMSketch(width, depth);
+
+    if(RedisModule_ModuleTypeSetValue(*key, CMSketchType, *cms) == REDISMODULE_ERR) {
+        return REDISMODULE_ERR;
+    }
+
+    return REDISMODULE_OK;
+}
+
+static int parseCreateArgs(RedisModuleCtx *ctx, RedisModuleString **argv, int argc,
+                    long long *width, long long *depth) {
+    if(CompareCStringToRedisString("cms.initbydim", argv[0]) == 1) {
+        if((RedisModule_StringToLongLong(argv[2], width) != REDISMODULE_OK)) 
+            return RedisModule_ReplyWithError(ctx, "CMS: invalid value");
+        if((RedisModule_StringToLongLong(argv[3], depth) != REDISMODULE_OK)) 
+            return RedisModule_ReplyWithError(ctx, "CMS: invalid value");        
+    } else {
+        double err, prob;
+        if((RedisModule_StringToDouble(argv[2], &err) != REDISMODULE_OK)) 
+            return RedisModule_ReplyWithError(ctx, "CMS: invalid value");
+        if((RedisModule_StringToDouble(argv[3], &prob) != REDISMODULE_OK)) 
+            return RedisModule_ReplyWithError(ctx, "CMS: invalid value");  
+
+        *width = ceil(2 / err);
+        *depth = ceil(log10f(prob) / log10f(0.5));
+    }
+    return REDISMODULE_OK;
+}
+
+int CMSketch_create(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    if(argc != 4)
+        return RedisModule_WrongArity(ctx);
+
+    CMSketch *cms;
+    long long width, depth;
+    RedisModuleString *keyName = argv[1];
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ|REDISMODULE_WRITE);
+
+    if(parseCreateArgs(ctx, argv, argc, &width, &depth) != REDISMODULE_OK)
+        return REDISMODULE_ERR;
+
+    if(RedisModule_KeyType(key) != REDISMODULE_KEYTYPE_EMPTY) {
+        RedisModule_CloseKey(key);
+        return RedisModule_ReplyWithError(ctx,"CMS: key already exists");
+    }
+
+    CreateCmsKey(ctx, keyName, width, depth, &cms, &key);
+    RedisModule_CloseKey(key);
+
+    RedisModule_Log(ctx, "info", "created new count-min sketch");
```

Comment:
This is being "big head" - I like that :) If that is the case, why not include more details like params and mebbe key name? That said, I'm not sure that the log message is really needed as we don't usually log key lifetime events.

Reply:
I need to learn more about it.
it was copied from reference w/o enough understanding of why and when.

Reply:
Removed for now